### PR TITLE
Backport same PRs as done in bitcoin#11445

### PR DESCRIPTION
**Pull request by codablock**
This PR backports the same PRs that were originally backported in bitcoin#11445. All PRs are backported from the original PRs and not from the commits found in bitcoin#11445. This ensures that we don't miss something.

### CODE REVIEW

File: test/functional/bitcoin_cli.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test bitcoin-cli"""
```
Comment (UdjinM6) [inline test/functional/bitcoin_cli.py]: Dashify `bitcoin-cli`? (whole file)

File: test/functional/test_framework/test_node.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    def node_encrypt_wallet(self, passphrase):
+        """"Encrypts the wallet.
+
+        This causes dashd to shutdown, so this method takes
+        care of cleaning up resources."""
+        self.encryptwallet(passphrase)
+        self.wait_until_stopped()
+
+class TestNodeCLI():
+    """Interface to bitcoin-cli for an individual node"""
```
Comment (UdjinM6) [inline test/functional/test_framework/test_node.py]: Dashify `bitcoin-cli`? (here and till the eof)

File: test/functional/test_framework/test_framework.py [hunk 3]
```diff
@@ -222,81 +185,121 @@ def main(self):
...
             binary = [None] * num_nodes
         assert_equal(len(extra_args), num_nodes)
         assert_equal(len(binary), num_nodes)
-        rpcs = []
+        old_num_nodes = len(self.nodes)
+        for i in range(num_nodes):
+            self.nodes.append(TestNode(old_num_nodes + i, self.options.tmpdir, extra_args[i], rpchost, timewait=timewait, binary=binary[i], stderr=stderr, mocktime=self.mocktime, coverage_dir=self.options.coveragedir))
+
+    def start_node(self, i, extra_args=None, stderr=None):
+        """Start a bitcoind"""
```
Comment (UdjinM6) [inline test/functional/test_framework/test_framework.py]:
```suggestion
        """Start a dashd"""
```

File: src/test/script_standard_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+    s.clear();
+    s << OP_1 << OP_1 << OP_CHECKMULTISIG;
+    BOOST_CHECK(!Solver(s, whichType, solutions));
+
+    // TX_NULL_DATA with other opcodes
+    s.clear();
+    s << OP_RETURN << std::vector<unsigned char>({75}) << OP_ADD;
+    BOOST_CHECK(!Solver(s, whichType, solutions));
+
+    // TX_WITNESS with unknown version
```
Comment (UdjinM6) [inline src/test/script_standard_tests.cpp]: Re-reviewing and noticed this...

File: src/test/script_standard_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+    s.clear();
+    s << OP_1 << ToByteVector(pubkey);
+    BOOST_CHECK(!Solver(s, whichType, solutions));
+
+    // TX_WITNESS with incorrect program size
```
Comment (UdjinM6) [inline src/test/script_standard_tests.cpp]: ...and this

File: src/test/script_standard_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+    s.clear();
+    s << OP_1 << ToByteVector(pubkey) << OP_1 << OP_CHECKMULTISIG;
+    BOOST_CHECK(!ExtractDestination(s, address));
+
+    // TX_NULL_DATA
+    s.clear();
+    s << OP_RETURN << std::vector<unsigned char>({75});
+    BOOST_CHECK(!ExtractDestination(s, address));
+
+    // TX_WITNESS_V0_KEYHASH
```
Comment (UdjinM6) [inline src/test/script_standard_tests.cpp]: same

File: src/test/script_standard_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    s.clear();
+    s << OP_0 << ToByteVector(pubkey);
+    BOOST_CHECK(!ExtractDestination(s, address));
+
+    // TX_WITNESS_V0_SCRIPTHASH
```
Comment (UdjinM6) [inline src/test/script_standard_tests.cpp]: same

File: src/test/script_standard_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+                *boost::get<CKeyID>(&addresses[0]) == pubkeys[0].GetID());
+    BOOST_CHECK(boost::get<CKeyID>(&addresses[1]) &&
+                *boost::get<CKeyID>(&addresses[1]) == pubkeys[1].GetID());
+
+    // TX_NULL_DATA
+    s.clear();
+    s << OP_RETURN << std::vector<unsigned char>({75});
+    BOOST_CHECK(!ExtractDestinations(s, whichType, addresses, nRequired));
+
+    // TX_WITNESS_V0_KEYHASH
```
Comment (UdjinM6) [inline src/test/script_standard_tests.cpp]: same

File: src/test/script_standard_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+    s.clear();
+    s << OP_0 << ToByteVector(pubkeys[0].GetID());
+    BOOST_CHECK(!ExtractDestinations(s, whichType, addresses, nRequired));
+
+    // TX_WITNESS_V0_SCRIPTHASH
```
Comment (UdjinM6) [inline src/test/script_standard_tests.cpp]: same

**Comment by codablock** [comment]
@UdjinM6 dropped the TX_WITNESS tests

File: test/functional/zmq_test.py [hunk 10]
```diff
@@ -7,12 +7,18 @@
...
+from codecs import encode
+
+from test_framework.mininode import dashhash
 from test_framework.test_framework import BitcoinTestFramework, SkipTest
 from test_framework.util import (assert_equal,
                                  bytes_to_hex_str,
                                  hash256,
                                 )
 
+def dashhash_helper(b):
```
Comment (nmarley) [inline test/functional/zmq_test.py]: I would probably have put a descriptive comment for this func here. Not immediately clear unless you understand what is going on. `# Accepts raw 80-byte block header, runs Dash PoW hash and reverses the bytes per convention` or something like that. But not a huge deal.

**Pull request merged**